pkg/execution/state/redis_state/cluster.go

```diff
@@ -0,0 +1,107 @@
+package redis_state
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"github.com/redis/rueidis"
+	"runtime"
+	"time"
+)
+
+type RetriableClient interface {
+	Do(ctx context.Context, cmd func(client rueidis.Client) rueidis.Completed) (resp rueidis.RedisResult)
+}
+
+type noopRetriableClient struct {
+	r rueidis.Client
+}
+
+func (r noopRetriableClient) B() rueidis.Builder {
+	return r.r.B()
+}
+
+func (r noopRetriableClient) Do(ctx context.Context, cmd func(client rueidis.Client) rueidis.Completed) (resp rueidis.RedisResult) {
+	return r.r.Do(ctx, cmd(r.r))
+}
+
+func newNoopRetriableClient(client rueidis.Client) RetriableClient {
+	return noopRetriableClient{client}
+}
+
+type retryClusterDownClient struct {
+	r rueidis.Client
+}
+
+func (r retryClusterDownClient) B() rueidis.Builder {
+	return r.r.B()
+}
+
+func (r retryClusterDownClient) do(ctx context.Context, cmd func(client rueidis.Client) rueidis.Completed, attempts int) rueidis.RedisResult {
+	resp := r.r.Do(ctx, cmd(r.r))
+
+	if err := resp.Error(); err != nil {
+		if ret, ok := rueidis.IsRedisErr(err); ok {
+			// retry on CLUSTERDOWN (in case we're scaling up/down)
+			if ret.IsClusterDown() {
+				if attempts == 5 {
+					return resp
+				}
+
+				time.Sleep(100 * time.Millisecond)
+				return r.do(ctx, cmd, attempts+1)
+			}
+		}
+	}
+
+	return resp
+}
+
+func (r retryClusterDownClient) Do(ctx context.Context, cmd func(client rueidis.Client) rueidis.Completed) (resp rueidis.RedisResult) {
+	return r.do(ctx, cmd, 0)
+}
+
+func newRetryClusterDownClient(r rueidis.Client) RetriableClient {
+	return &retryClusterDownClient{r: r}
+}
+
+// NewClusterLuaScript creates a Lua instance whose Lua.Exec uses EVALSHA and EVAL.
+func NewClusterLuaScript(script string) *RetriableLua {
```

Comment:
We also need to introduce a forked version of Rueidis Lua scripts that works with our retriable client.